internal/ext/stringsx/stringsx.go

```diff
@@ -17,12 +17,37 @@ package stringsx
 
 import (
 	"strings"
+	"unicode/utf8"
 
 	"github.com/bufbuild/protocompile/internal/ext/iterx"
+	"github.com/bufbuild/protocompile/internal/ext/slicesx"
 	"github.com/bufbuild/protocompile/internal/ext/unsafex"
 	"github.com/bufbuild/protocompile/internal/iter"
 )
 
+// Rune returns the rune at the given index.
+//
+// Returns 0, false if out of bounds. Returns U+FFFD, false if rune decoding fails.
+func Rune[I slicesx.SliceIndex](s string, idx I) (rune, bool) {
+	if !slicesx.BoundsCheck(idx, len(s)) {
+		return 0, false
+	}
+	r, _ := utf8.DecodeRuneInString(s[idx:])
+	return r, r != utf8.RuneError
+}
+
+// Rune returns the previous rune at the given index.
```

Comment:
Ditto above. So maybe something like "rune that ends just before given byte index".